module/zfs/dsl_dataset.c

```diff
@@ -1852,6 +1882,11 @@ dsl_dataset_snapshot_tmp_check(void *arg, dmu_tx_t *tx)
 		dsl_dataset_rele(ds, FTAG);
 		return (error);
 	}
+	error = dsl_dataset_snapshot_reserve_space(ds, tx);
+	if (error != 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (error);
+	}
```

Comment:
Shouldn't this call be moved to the end after `dsl_dataset_user_hold_check_one()`?  Otherwise, we can potentially hit the same issue if one of the calls subsequent to `dsl_dataset_snapshot_reserve_space()` fails.